### PR TITLE
Support per-mod variable modification limits

### PR DESCRIPTION
**Pull request by pgarrett-scripps**
## Summary

Adds optional per-modification occurrence limits and a cap on the total number
of variable-modification combinations generated for each peptide.

This implements the structured variable-modification format proposed in #231:

```json
"variable_mods": {
  "M": [
    {
      "mass": 15.9949,
      "max_count": 1
    }
  ]
}
```

Bare mass values remain supported for backward compatibility:

```json
"variable_mods": {
  "M": [15.9949]
}
```

The two formats can also be mixed within the same configuration.

## Changes

- Add structured variable-modification entries containing `mass` and optional `max_count`
- Add optional database-level `max_combinations`
- Preserve the existing `max_variable_mods` behavior
- Prefer variants with fewer modifications when applying the combination cap
- Preserve structured entries when serializing parameters to `results.json`
- Update documentation and changelog
- Add tests covering parsing, compatibility, limits, caps, and edge cases

`max_combinations` includes the unmodified peptide. A value of `0` is normalized
to `1`, matching the existing behavior of `max_variable_mods`.

Names, labile modifications, and neutral-loss configuration are intentionally
left for future work.

## Backward compatibility

Existing configurations using arrays of numeric masses continue to deserialize
and behave as before. When omitted:

- `max_count` remains unlimited
- `max_combinations` remains unlimited
- `max_variable_mods` retains its existing default and behavior

## Validation

- `cargo fmt --all -- --check`
- `cargo check --workspace --all-targets`
- `cargo test --workspace` — 108 tests passed
- Release build
- End-to-end CLI searches using both structured and legacy configurations
- Verified structured configuration round-trips through `results.json`

Addresses #231.

### CODE REVIEW

**Comment by lazear** [comment]
I will try to review next week!

File: crates/sage/src/database.rs [hunk 1]
```diff
@@ -74,10 +74,16 @@ pub struct Builder {
...
+    /// Variable modifications to add to matching amino acids.
+    /// Each entry is either a bare mass (`15.9949`) or an object with `mass` and
+    /// optional `max_count` fields (`{"mass": 15.9949, "max_count": 1}`).
+    pub variable_mods: Option<HashMap<String, Vec<VarModEntry>>>,
     /// Limit number of variable modifications on a peptide
     pub max_variable_mods: Option<usize>,
+    /// Hard cap on the total peptide variants generated per input peptide,
+    /// including its unmodified form. Values below 1 are normalized to 1.
+    /// Variants with fewer PTMs are preferred (generated first).
+    pub max_combinations: Option<usize>,
```
Comment (lazear) [inline crates/sage/src/database.rs:86]: Not sure how I feel about this - I guess it's optional? My gut feeling is that "silently" truncating variants is probably bad

**Review by lazear — COMMENTED** [review]
LGTM overall, only question is whether the extra machinery to cap max peptide species is necessary!

File: crates/sage/src/database.rs (hunk 1 above)
Comment (pgarrett-scripps) [reply crates/sage/src/database.rs:86]: Truncation is necessary in some cases. The number of variants can scale exponentially, and in my experience a select few peptides, especially in histone searches, blow up enough to make the search infeasible without a limit. It's opt-in and defaults to unlimited, so existing searches don't change. I agree silent truncation isn't ideal, though, so I'm happy to add a warning that reports how many peptides hit the cap.
